### PR TITLE
ValueTracking: Merge fcmpImpliesClass and fcmpToClassTest

### PR DESCRIPTION
**Pull request by arsenm**
Rushing this one out before vacation starts. Refactoring on top of #66505

### CODE REVIEW

**Comment by dtcxzyw** [comment]
Is there any progress?

**Review by spavloff — APPROVED** [review]
LGTM.

File: llvm/test/Transforms/Attributor/nofpclass-implied-by-fcmp.ll [hunk 1]
```diff
@@ -2641,8 +2641,8 @@ define float @assume_false_smallest_normal(float %arg) {
 }
 
 define float @clamp_false_nan(float %arg) {
-; CHECK-LABEL: define float @clamp_false_nan(
-; CHECK-SAME: float returned [[ARG:%.*]]) #[[ATTR2]] {
+; CHECK-LABEL: define nofpclass(nan inf nzero sub norm) float @clamp_false_nan(
+; CHECK-SAME: float returned nofpclass(nan inf nzero sub norm) [[ARG:%.*]]) #[[ATTR2]] {
```
Comment (arsenm) [inline llvm/test/Transforms/Attributor/nofpclass-implied-by-fcmp.ll:2645]: @jdoerfert this is the case I was looking at in #78507. It makes no sense to infer anything here

Comment (jdoerfert) [reply llvm/test/Transforms/Attributor/nofpclass-implied-by-fcmp.ll:2645]: I know what is wrong, did https://github.com/jdoerfert/llvm-project/tree/fpclass_fixes help?
It is just very intrusive and you probably want to populate the switch.

Comment (arsenm) [reply llvm/test/Transforms/Attributor/nofpclass-implied-by-fcmp.ll:2645]: Think it's the first part, not the second

Comment (arsenm) [reply llvm/test/Transforms/Attributor/nofpclass-implied-by-fcmp.ll:2645]: This turned out to be the issue fixed in 1a6953a75d3120a4f9196911e6a6b79fcc315553

**Pull request merged**